wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/ApiConnection.java

```diff
@@ -0,0 +1,510 @@
+	 * @throws LoginFailedException
+	 */
+	void confirmLogin(String token, String username, String password)
+			throws IOException, LoginFailedException {
+		Map<String, String> params = new HashMap<String, String>();
+		params.put("action", "login");
+		params.put("lgname", username);
+		params.put("lgpassword", password);
+		params.put("lgtoken", token);
+		params.put("format", "json");
+
+		JsonNode root = this.mapper.readTree(sendRequest("POST", params));
+
+		String result = root.get("login").get("result").textValue();
+		if (result.equals(ApiConnection.LOGIN_RESULT_SUCCESS)) {
+			this.loggedIn = true;
+			this.username = username;
+			this.password = password;
+			return;
+		} else {
+			String message = getLoginErrorMessage(result);
+			logger.warn(message);
+			throw new LoginFailedException(message);
+		}
+	}
+
+	/**
+	 * Returns a user-readable message for a given API response.
+	 *
+	 * @param loginResult
+	 *            a API login request result string other than
+	 *            {@link #LOGIN_RESULT_SUCCESS}
+	 * @return error message
+	 */
+	String getLoginErrorMessage(String loginResult) {
+		switch (loginResult) {
+		case ApiConnection.LOGIN_WRONG_PASS:
+			return loginResult + ": Wrong Password.";
+		case ApiConnection.LOGIN_WRONG_PLUGIN_PASS:
+			return loginResult
+					+ ": Wrong Password. An authentication plugin rejected the password.";
+		case ApiConnection.LOGIN_NOT_EXISTS:
+			return loginResult + ": Username does not exist.";
+		case ApiConnection.LOGIN_BLOCKED:
+			return loginResult + ": User is blocked.";
+		case ApiConnection.LOGIN_EMPTY_PASS:
+			return loginResult + ": Password is empty.";
+		case ApiConnection.LOGIN_NO_NAME:
+			return loginResult + ": No user name given.";
+		case ApiConnection.LOGIN_CREATE_BLOCKED:
+			return loginResult
+					+ ": The wiki tried to automatically create a new account for you,"
+					+ "but your IP address has been blocked from account creation.";
+		case ApiConnection.LOGIN_ILLEGAL:
+			return loginResult + ": Username is illegal.";
+		case ApiConnection.LOGIN_THROTTLED:
+			return loginResult + ": Too many login attempts in a short time.";
+		case ApiConnection.LOGIN_WRONG_TOKEN:
+			return loginResult + ": Token is wrong.";
+		case ApiConnection.LOGIN_NEEDTOKEN:
+			return loginResult + ": Token or session ID is missing.";
+		default:
+			return "Login Error: " + loginResult;
+		}
+	}
+
+	/**
+	 * Logs the current user out.
+	 *
+	 * @throws IOException
+	 */
+	public void logout() throws IOException {
+		if (this.loggedIn) {
+			Map<String, String> params = new HashMap<String, String>();
+			params.put("action", "logout");
+			params.put("format", "json"); // reduce the output
+			this.sendRequest("POST", params);
+			this.loggedIn = false;
+			this.username = "";
+			this.password = "";
+		}
+	}
+
+	/**
+	 * Reads out the Set-Cookie Header Fields and fills the cookie map of the
+	 * API connection with it.
+	 * <p>
+	 * TODO Why does this method clear all existing cookies?
```

Comment:
I do not understand why the fillCookies method deletes all existing cookies. This seems to be a potential source of errors. I would expect that cookies can always be overwritten by sending a new cookie with a given key.